src/statics.rs

```diff
@@ -80,29 +87,98 @@ pub(crate) static EVENT_METADATA: once_cell::sync::Lazy<
+                        // TODO: This only works when comparing the static entries generated by the logging macros.
+                        let lhs = &self.meta.identity as *const tracing_core::callsite::Identifier as usize;
+                        let rhs = &other.meta.identity as *const tracing_core::callsite::Identifier as usize;
+                        lhs.partial_cmp(&rhs)
+                    }
+                }
+            },
+            Some(x) => Some(x)
+        }
     }
 }
 
 impl core::cmp::Ord for crate::_details::ParsedEventMetadata {
     fn cmp(&self, other: &Self) -> std::cmp::Ordering {
-        self.identity_hash.cmp(&other.identity_hash)
+        match self.identity_hash.cmp(&other.identity_hash) {
+            cmp::Ordering::Equal => {
+                match self.meta.identity == other.meta.identity {
+                    true => cmp::Ordering::Equal,
+                    false => {
+                        // We need to do *something* to sort two different callsites that hit a hash collision.
+                        // TODO: This only works when comparing the static entries generated by the logging macros.
+                        let lhs = &self.meta.identity as *const tracing_core::callsite::Identifier as usize;
+                        let rhs = &other.meta.identity as *const tracing_core::callsite::Identifier as usize;
+                        lhs.cmp(&rhs)
+                    }
+                }
+            },
+            x => x
+        }
     }
 }
 
-pub(crate) fn get_event_metadata(id: &tracing::callsite::Identifier) -> Option<&'static crate::_details::ParsedEventMetadata> {
-    let hash = BH.hash_one(id);
-    let etw_meta = EVENT_METADATA.binary_search_by_key(&hash, |m| { m.identity_hash });
-    match etw_meta {
-        Ok(idx) => Some(&EVENT_METADATA[idx]),
-        _ => None
+pub(crate) fn get_event_metadata(id: &tracing::callsite::Identifier) -> Option<&'static crate::_details::EventMetadata> {
+    let bh = FnvHasher::default();
+    let identity_hash = bh.hash_one(id);
+    let idx = EVENT_METADATA.partition_point(|other| other.identity_hash > identity_hash);
+    let mut cur = idx;
+    while cur <EVENT_METADATA.len() {
+        let meta = &EVENT_METADATA[cur];
+        if meta.identity_hash == identity_hash {
+            if meta.meta.identity == *id {
+                return Some(meta.meta);
+            }
+        }
+        else {
+            break;
+        }
+        cur += 1;
     }
+    None
+}
+
+pub(crate) struct EventMetadataEnumerator {
+    current_index: usize
+}
+
+impl FusedIterator for EventMetadataEnumerator {}
+
+impl Iterator for EventMetadataEnumerator {
+    type Item = &'static EventMetadata;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current_index >= EVENT_METADATA.len() {
+            return None;
+        }
+
+        self.current_index += 1;

```

Comment:
This still allows for indexing off the end of the array.

Start condition: current_index == len - 1.

`current_index >= len` is false, so we don't exit early.

`current_index += 1` makes current_index == len.

`EVENT_METADATA[current_index]` is now `EVENT_METADATA[len]` which is past the end of the array.